2_BinaryTree/2_BinaryTree/BinaryTree.cpp

```diff
@@ -0,0 +1,451 @@
+}
+
+void BinaryTree::clear()
+{
+    clearFrom(m_root);
+    delete m_root;
+    m_root = nullptr;
+}
+
+BinaryTree BinaryTree::clone() const
+{
+    return clone(m_root);
+}
+
+BinaryTree BinaryTree::clone(Node* root) const
+{
+    if (!root)
+        return *this;
+    BinaryTree cloneTree;
+    cloneTree.m_root = _clone(root);
+    return cloneTree;
+}
+
+BinaryTree::Node* BinaryTree::_clone() const
+{
+    return _clone(m_root);
+}
+
+BinaryTree::Node* BinaryTree::_clone(Node* root) const
+{
+    Node* cloneRoot = nullptr;
+    if (!root)
+        return cloneRoot;
+    cloneRoot = new Node(root->getKey());
+    cloneRoot->setLeft(_clone(root->getLeft()));
+    cloneRoot->setRight(_clone(root->getRight()));
+    return cloneRoot;
+}
+
+void BinaryTree::_treeToList(std::list<Node*>& nodeList) const
+{
+    Node* current = nullptr;
+    std::list<Node*> tmp;
+    tmp = nodeList;
+    while (!tmp.empty())
+    {
+        current = tmp.front();
+        tmp.pop_front();
+        if (current->getLeft())
+        {
+            nodeList.push_back(current->getLeft());
+            tmp.push_back(current->getLeft());
+        }
+        if (current->getRight())
+        {
+            nodeList.push_back(current->getRight());
+            tmp.push_back(current->getRight());
+        }
+    }
+}
+
+bool BinaryTree::isBalanced() const
+{
+    return isBalanced(m_root);
+}
+
+bool BinaryTree::isBalanced(Node* root) const
+{
+    if (!root)
+        return true;
+    int left = height(root->getLeft());
+    int right = height(root->getRight());
+    int difference = right - left;
+    bool isBalance = false;
+    if (difference == 0 || difference == 1 || difference == -1)
+        isBalance = true;
+    return isBalance * isBalanced(root->getLeft()) * isBalanced(root->getRight());
+}
+
+bool BinaryTree::isEmpty() const
+{
+    return m_root == nullptr;
+}
+
+int BinaryTree::nodeCount() const
+{
+    int result = 0;
+    if (!m_root)
+        return result;
+    std::list<Node*> listNode;
+    listNode.push_back(m_root);
+    _treeToList(listNode);
+    result = listNode.size();
+    return result;
+}
+
+int BinaryTree::height() const
+{
+    return height(m_root);
+}
+
+int BinaryTree::height(Node* root) const
+{
+    if (!root)
+        return 0;
+    else
+    {
+        int leftHeight = height(root->getLeft());
+        int rightHeight = height(root->getRight());
+        return std::max(leftHeight, rightHeight) + 1;
+    }
+}
+
+BinaryTree::Node* BinaryTree::find(const int key) const
+{
+    return find(m_root, key);
+}
+
+BinaryTree::Node* BinaryTree::find(Node* root, const int key) const
+{
+    if (!root)
+        return nullptr;
+
+    if (root->getKey() == key)
+        return root;
+    else
+    {
+        Node* left = find(root->getLeft(), key);
+        Node* right = find(root->getRight(), key);
+        if (left)
+            return left;
+        else if (right)
+            return right;
+        else
+            return nullptr;

```

Comment:
Зачем сразу искать в левом поддереве, если мы уже могли найти узел в правом?
```suggestion
        Node* node = find(root->getLeft(), key);
        if (!node)
            node = find(root->getRight(), key);
        return node;
```

Reply:
сделано